docs/articles/documentation/extending-testcafe/custom-browser-provider-plugin/index.md

```diff
@@ -0,0 +1,180 @@
+
+```bash
+npm install -g yo
+npm install -g generator-testcafe-browser-provider
+```
+
+Create a new directory where the generator should place your scaffolded project files and go into it.
+
+```bash
+mkdir my-browser-provider
+cd my-browser-provider
+```
+
+> It is recommended to name the directory as you would like to name the browser provider project. When you run the generator, it will automatically suggest the browser provider name that matches the browser provider directory name. As for the provider package name, it will consist of two parts - the `{testcafe-browser-provider-}` prefix and the name of the provider itself, for example, `{testcafe-browser-provider-my-provider}`.
+
+Then run the browser provider generator to create a new project:
+
+```bash
+yo testcafe-browser-provider
+```
+
+The generator will provide additional settings to customize your browser provider.
+Then Yeoman will automatically scaffold out your browser provider, install the required dependencies, and pull in several useful Gulp tasks for your workflow.
+
+## Implementing the Browser Provider
+
+Once the browser provider has been scaffolded out, go to the provider directory and open the `src/index.js` file.
+
+There are several required and optional provider methods you can implement.
+
+```js
+// Required methods
+async openBrowser (/* id, browserInfo, startPage */) {
+    throw new Error('Not implemented!');
+},
+
+async closeBrowser (/* id, pageInfo */) {
+    throw new Error('Not implemented!');
+},
+
+// Optional initialization and cleanup methods
+async init () {
+    return;
+},
+
+async dispose () {
+    return;
+},
+
+// Optional methods for multi-browser support
+async getBrowserList () {
+    throw new Error('Not implemented!');
+},
+
+async isValidBrowserName (/* browserName */) {
+    return true;
+},
+
+
+// Optional extra methods
+async resizeWindow (/* id, pageInfo, width, height */) {
+    this.reportWarning('The screenshot functionality is not supported by the "<%= providerName %>" browser provider.');
+},
+
+async takeScreenshot (/* id, pageInfo, screenshotPath */) {
+    this.reportWarning('The window resize functionality is not supported by the "<%= providerName %>" browser provider.');
+}
+```
+
+For details, see [Browser Provider Methods](browser-provider-methods.md).
+
+### Example
+
+The following example code demonstrates how you can implement the simple browser provider for portable browsers.
+
+```js
+var exec = require('child_process').exec;
+var browserTools = require('testcafe-browser-tools');
+
+export default {
+    isMultiBrowser: true
+
+    async openBrowser (id, browserName, startPage) {
+        switch (browserName) {
+            case 'firefox':
+                exec('C:\FirefoxPortable\firefox.exe --new-window ' + startPage);
+                break;
+            case 'chrome':
+                exec('C:\ChromePortable\chrome.exe --new-window ' + startPage);
```

Comment:
It is `async` because it must be so to match required interface, if we are talking about that concrete provider. Interface is `async`, because we must search installed browsers, or create remote connection to Saucelabs, or because `phantomjs` API is `async` too.